src/facade/dragonfly_connection.cc

```diff
@@ -102,18 +102,83 @@ struct Connection::Shutdown {
 };
 
 struct Connection::Request {
-  absl::FixedArray<MutableSlice, 6> args;
+  struct PipelineMsg {
+    absl::FixedArray<MutableSlice, 6> args;
 
-  // I do not use mi_heap_t explicitly but mi_stl_allocator at the end does the same job
-  // of using the thread's heap.
-  // The capacity is chosen so that we allocate a fully utilized (256 bytes) block.
-  absl::FixedArray<char, kReqStorageSize, mi_stl_allocator<char>> storage;
-  AsyncMsg* async_msg = nullptr;  // allocated and released via mi_malloc.
+    // I do not use mi_heap_t explicitly but mi_stl_allocator at the end does the same job
+    // of using the thread's heap.
+    // The capacity is chosen so that we allocate a fully utilized (256 bytes) block.
+    absl::FixedArray<char, kReqStorageSize, mi_stl_allocator<char>> storage;
 
-  Request(size_t nargs, size_t capacity) : args(nargs), storage(capacity) {
+    PipelineMsg(size_t nargs, size_t capacity) : args(nargs), storage(capacity) {
+    }
+  };
+
+  struct PubSubMsg {
+    AsyncMsg* async_msg = nullptr;  // allocated and released via mi_malloc.
+
+    PubSubMsg(const PubMessage& pub_msg, fibers_ext::BlockingCounter bc)
+        : async_msg{(AsyncMsg*)mi_malloc(sizeof(AsyncMsg))} {
+      async_msg->bc = std::move(bc);
+      async_msg->pub_msg = pub_msg;
+    }
+
+    void Free() {
```

Comment:
since this class wraps a pod type I do not see any problem to declare a c++ d'tor here.  Free method is not needed imho.

Reply:
you might need to make it moveable only to work.

Reply:
This was because it don't have trivial memory allocation. We are not allocating from the "normal" heap. It turned out that using the dtor cause it to free the memory when the object was place in the variant. see the ctor for details about the memory management. The original code use the dtor and it wasn't good.

Reply:
See my comment about - maybe we don't need to allocate at all in this case (other than the Request object itself)

Reply:
yes, I think we can avoid the allocation of AsyncMsg and use it in variant directly.